nemo/collections/nlp/parts/nlp_overrides.py

```diff
@@ -0,0 +1,233 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+from typing import Any, Dict, List, Optional, Union
+
+import torch
+from megatron import mpu
+from megatron.checkpointing import get_checkpoint_version, set_checkpoint_version
+from megatron.initialize import _set_random_seed
+from pytorch_lightning.core.lightning import LightningModule
+from pytorch_lightning.overrides import LightningDistributedModule
+from pytorch_lightning.plugins.environments.cluster_environment import ClusterEnvironment
+from pytorch_lightning.plugins.training_type.ddp import DDPPlugin
+from pytorch_lightning.trainer.connectors.checkpoint_connector import CheckpointConnector
+from pytorch_lightning.utilities import rank_zero_warn
+from pytorch_lightning.utilities.cloud_io import atomic_save
+from torch.nn.parallel import DistributedDataParallel
+
+from nemo.collections.nlp.modules.common.megatron.megatron_bert import MegatronBertEncoder
+from nemo.utils import AppState, logging
+
+
+class NLPDDPPlugin(DDPPlugin):
+    """ DDP plugin for Pytorch Lightning. Needed to customize DDP for model parallel models.
+    """
+
+    distributed_backend = "ddp"
+
+    def __init__(
+        self,
+        parallel_devices: Optional[List[torch.device]] = None,
+        num_nodes: int = 1,
+        cluster_environment: ClusterEnvironment = None,
+        sync_batchnorm: bool = False,
+        **kwargs: Union[Any, Dict[str, Any]],
+    ) -> None:
+        super().__init__(parallel_devices, num_nodes, cluster_environment, sync_batchnorm, **kwargs)
+
+    def init_ddp_connection(self, global_rank: int, world_size: int) -> None:
+        # call PTL init ddp
+        super().init_ddp_connection(global_rank, world_size)
+
+        # init model parallel
+        app_state = AppState()
+
+        if app_state.model_parallel_size is not None:
+
+            if isinstance(self.lightning_module.bert_model, MegatronBertEncoder):
+
+                if app_state.model_parallel_group is None:
+                    self.init_model_parallel(app_state.global_rank, app_state.world_size)
+
+    def start_training(self, trainer: 'Trainer') -> None:
+        """ PTL Hook that is called after DPP is initialized. """
+
+        if isinstance(self.lightning_module.bert_model, MegatronBertEncoder):
+            app_state = AppState()
+            if app_state.model_parallel_size is not None:
+                # mpu grad clipping needs parameters to have the attribute model_parallel
+                parameters = self.lightning_module.parameters()
+                for p in parameters:
+                    if not hasattr(p, 'model_parallel'):
+                        p.model_parallel = False
+
+                # TODO: figure out how to override clip gradients again
+                # Update PTL trainer to use our _clip_gradients
+                # self._trainer.accelerator_backend._clip_gradients = self._clip_gradients
```

Comment:
Lets raise a user visible warning if they try to clip grads and its not implemented yet.

Reply:
Oh you can still clip grads just won't be over model parallel groups until we add that back.